lib/utils/metrics.js

```diff
@@ -0,0 +1,146 @@
+'use strict';
+
+const assert = require('assert');
+const uuid = require('uuid');
+
+/** @typedef { import('bmutex').Lock } Lock */
+
+const metrics = exports;
+
+/**
+ * @typedef {Object} Honey
+ */
+
+class Event {
+  /**
+   *
+   * @param {string} name the name for the event
+   * @param {Event | Honey} parent a parent Event or a libhoney instance
+   */
+  constructor(name, parent) {
+    assert(name, 'Event name is required');
+    assert(parent, 'Parent event or libhoney instance is required.');
+    if (parent instanceof Event) {
+      this.honey = parent.honey;
+      this.parent = parent;
+      this.trace_id = parent.trace_id;
+    } else {
+      this.honey = parent;
+      this.parent = null;
+      this.trace_id = uuid.v4();
+    }
+
+    this.event = this.honey.newEvent();
+
+    this.span_id = uuid.v4();
+    this.startTime = process.hrtime();
+    this.timers = new Map();
+
+    this.addField('name', name);
+    this.addField('service_name', 'hsd');
+    this.addField('trace.span_id', this.span_id);
+    this.addField('trace.trace_id', this.trace_id);
+    if(this.parent) {
+      this.addField('trace.parent_id', this.parent.span_id);
+      this.addField('parent_id', this.parent.span_id);
+    }
+    this.event.timestamp = Date.now();
+  }
+
+  startTimer(name) {
+    if(name) {
+      this.timers.set(name, process.hrtime());
+    } else {
+      this.startTime = process.hrtime();
+    }
+  }
+
+  /**
+   * Measure how long it takes to get a lock
+   * @param {Lock} lock a bmutex Lock
+   * @returns {Promise<any>} a function to call to release the lock
+   */
+  async measureLockWait(lock) {
+    this.addField('preLockQueueLength', lock.jobs.length);
+    const preLockTime = process.hrtime();
+    const unlock = lock.lock();
+
+    this.addField('postLockQueueLength', lock.jobs.length);
+    this.addField(
+      'lockWaitTime',
+      metrics.timerToMs(process.hrtime(preLockTime))
+    );
+
+    return unlock;
+  }
+
+  stopTimer(name) {
+    const start = this.timers.get(name);
+    if (start) {
+      this.timers.delete(name);
+      this.event.addField(
+        `timer:${name}`,
+        metrics.timerToMs(process.hrtime(start))
+      );
+    }
+  }
+
+  time(timerName, cb) {
+    this.startTimer(timerName);
+    const result = cb();
+    this.stopTimer(timerName);
+    return result;
+  }
+
+  async asyncSubEvent(name, cb) {
+    const subEvent = metrics.newEvent(name, this);
+    try {
+      await cb();
```

Comment:
I haven't scanned everything this is used yet, but it seems like a good idea to `return await cb()` so that this could be drop-in replaced whenever we do cb(). Or is the intended usage different and this definitely isn't necessary?

Reply:
It should, yes